proof/hash_to_g1.saw

```diff
@@ -0,0 +1,490 @@
+  llvm_execute_func [p_ptr, msg_ptr, llvm_term {{ `msg_len : [64] }}, DST_ptr, llvm_term {{ `DST_len : [64] }}, aug_ptr, llvm_term {{ `aug_len : [64] }}];
+  llvm_points_to p_ptr (llvm_term {{ (POINTonE1_rep (hash_to_curve_opt_impl (aug#msg, DST))) }});
+};
+
+let map_to_g1_impl_nonnull_spec = do {
+  out_ptr <- llvm_alloc POINTonE1_type;
+  (u,u_ptr) <- ptr_to_fresh_readonly "u" vec384_type;
+  (v,v_ptr) <- ptr_to_fresh_readonly "v" vec384_type;
+  llvm_precond {{ fp_invariant u }};
+  llvm_precond {{ fp_invariant v }};
+  llvm_execute_func [out_ptr, u_ptr, v_ptr];
+  llvm_points_to out_ptr (llvm_term {{ POINTonE1_rep (map_to_g1_impl [fp_abs u, fp_abs v]) }});
+};
+
+let isogeny_map_to_E1_impl_alias_spec = do {
+  (out, out_ptr) <- ptr_to_fresh "out" POINTonE1_type;
+  llvm_precond {{ POINTonE1_invariant out }};
+  llvm_execute_func [out_ptr, out_ptr];
+  llvm_points_to out_ptr (llvm_term {{ POINTonE1_rep (isogeny_map_to_E1_impl (POINTonE1_abs out)) }} );
+};
+
+let map_to_isogenous_E1_impl_spec = do {
+  out_ptr <- llvm_alloc POINTonE1_type;
+  (u, u_ptr) <- ptr_to_fresh_readonly "u" vec384_type;
+  llvm_precond {{ fp_invariant u }};
+  llvm_execute_func [out_ptr, u_ptr];
+  llvm_points_to out_ptr (llvm_term {{ POINTonE1_rep (map_to_curve_simple_swu_impl (fp_abs u)) }});
+};
+
+let map_to_g1_nonnull_spec = do {
+  out_ptr <- llvm_alloc POINTonE1_type;
+  (u,u_ptr) <- ptr_to_fresh_readonly "u" vec384_type;
+  (v,v_ptr) <- ptr_to_fresh_readonly "v" vec384_type;
+  llvm_precond {{ fp_invariant u }};
+  llvm_precond {{ fp_invariant v }};
+  llvm_execute_func [out_ptr, u_ptr, v_ptr];
+  llvm_points_to out_ptr (llvm_term {{ POINTonE1_rep (map_to_g1_impl [fp_abs u, fp_abs v]) }});
+};
+
+// other overrides:
+
+let dbl_n_add_spec n = do {
+  (out, out_ptr) <- ptr_to_fresh "out" POINTonE1_type;
+  (p, p_ptr) <- ptr_to_fresh_readonly "p" POINTonE1_type;
+  llvm_precond {{ POINTonE1_invariant out /\ POINTonE1_invariant p }};
+  llvm_execute_func [out_ptr, llvm_term {{`n:[64]}}, p_ptr];
+  llvm_points_to out_ptr (llvm_term {{ POINTonE1_rep (dbl_n_add (POINTonE1_abs out) `n (POINTonE1_abs p)) }});
+};
+
+// the code calls this assembly function:
+let sgn0x_pty_mont_384_spec = do {
+  (a, a_ptr) <- ptr_to_fresh_readonly "a" vec384_type;
+  p_ptr <- alloc_init_readonly vec384_type (llvm_term {{ vec384_rep (from_Fp `p) }});
+  llvm_execute_func [a_ptr, p_ptr, llvm_term {{ 0x89f3fffcfffcfffd }}];
+  llvm_return (llvm_term {{ zext`{m=64} [2*(fp_abs a)>`p, (from_Fp (fp_abs a))!0] }});
+};
+
+// we also need this:
+let recip_sqrt_fp_3mod4_spec = do {
+    out_ptr <- crucible_alloc vec384_type;
+    (inp, inp_ptr) <- ptr_to_fresh_readonly "inp" vec384_type;
+    crucible_precond {{ fp_invariant inp }};
+    crucible_execute_func [out_ptr, inp_ptr];
+    crucible_points_to out_ptr (crucible_term {{fp_rep (Fp.div (Fp.field_unit, sqrt_fp (fp_abs inp))) }});
+    };
+
+sgn0x_pty_mont_384_ov <- admit "sgn0x_pty_mont_384" sgn0x_pty_mont_384_spec;
```

Comment:
I thought you had proved this override

Reply:
Yes indeed

Reply:
Actually no, it's only been tested with quickcheck.